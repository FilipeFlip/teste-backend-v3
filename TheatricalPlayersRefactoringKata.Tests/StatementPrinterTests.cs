using System.Collections.Generic;
using System.Globalization;
using System.Threading.Tasks;
using ApprovalTests;
using ApprovalTests.Reporters;
using Microsoft.EntityFrameworkCore;
using TheatricalPlayersRefactoringKata.Data;
using TheatricalPlayersRefactoringKata.Services;
using Xunit;

namespace TheatricalPlayersRefactoringKata.Tests
{
    public class StatementPrinterTests
    {
        private readonly ApplicationDbContext _context;
        private readonly StatementPrinter _statementPrinter;

        public StatementPrinterTests()
        {
            var options = new DbContextOptionsBuilder<ApplicationDbContext>()
                .UseInMemoryDatabase(databaseName: "TestDatabase")
                .Options;

            _context = new ApplicationDbContext(options);

            var playCategories = new Dictionary<string, IPlayCategory>
            {
                { "tragedy", new TragedyCategory() },
                { "comedy", new ComedyCategory() },
                { "history", new HistoricalCategory() }
            };

            _statementPrinter = new StatementPrinter(_context, playCategories);

            SeedDatabase();
        }

        private void SeedDatabase()
        {
            var plays = new List<Play>
            {
                new Play { Title = "Hamlet", Category = "tragedy", Lines = 4024 },
                new Play { Title = "As You Like It", Category = "comedy", Lines = 2670 },
                new Play { Title = "Othello", Category = "tragedy", Lines = 3560 },
                new Play { Title = "Henry V", Category = "history", Lines = 3227 },
                new Play { Title = "King John", Category = "history", Lines = 2648 },
                new Play { Title = "Richard III", Category = "history", Lines = 3718 }
            };

            var invoice = new Invoice
            {
                CustomerName = "BigCo",
                Performances = new List<Performance>
                {
                    new Performance { Play = plays[0], Audience = 55 },
                    new Performance { Play = plays[1], Audience = 35 },
                    new Performance { Play = plays[2], Audience = 40 },
                    new Performance { Play = plays[3], Audience = 20 },
                    new Performance { Play = plays[4], Audience = 39 },
                    new Performance { Play = plays[5], Audience = 20 }
                }
            };

            _context.Plays.AddRange(plays);
            _context.Invoices.Add(invoice);
            _context.SaveChanges();
        }

        [Fact]
        [UseReporter(typeof(DiffReporter))]
        public async Task TestStatementExampleLegacy()
        {
            var invoice = await _context.Invoices
                .Include(i => i.Performances)
                .ThenInclude(p => p.Play)
                .FirstOrDefaultAsync(i => i.CustomerName == "BigCo");

            var result = await _statementPrinter.PrintStatementAsync(invoice.Id, CultureInfo.InvariantCulture);

            Approvals.Verify(result);
        }

        [Fact]
        [UseReporter(typeof(DiffReporter))]
        public async Task TestTextStatementExample()
        {
            var invoice = await _context.Invoices
                .Include(i => i.Performances)
                .ThenInclude(p => p.Play)
                .FirstOrDefaultAsync(i => i.CustomerName == "BigCo");

            var result = await _statementPrinter.PrintStatementAsync(invoice.Id, CultureInfo.InvariantCulture);

            Approvals.Verify(result);
        }
    }
}
