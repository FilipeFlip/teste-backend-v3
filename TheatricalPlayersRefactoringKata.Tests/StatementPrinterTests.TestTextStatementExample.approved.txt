Statement for BigCo
- Hamlet: R$ 350,00 (55 seats)
  Credits: 25
- As You Like It: R$ 380,00 (35 seats)
  Credits: 25
- Othello: R$ 200,00 (40 seats)
  Credits: 10
- Henry V: R$ 260,00 (20 seats)
  Credits: 0
- King John: R$ 602,00 (39 seats)
  Credits: 9
- Henry V: R$ 260,00 (20 seats)
  Credits: 0
Total amount: R$ 2.052,00
Total points: 69
